src/framework/uicomponents/view/selectmultipledirectoriesmodel.cpp

```diff
@@ -0,0 +1,184 @@
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "selectmultipledirectoriesmodel.h"
+
+#include "translation.h"
+
+#include "uicomponents/view/itemmultiselectionmodel.h"
+
+using namespace mu::uicomponents;
+
+SelectMultipleDirectoriesModel::SelectMultipleDirectoriesModel(QObject* parent)
+    : QAbstractListModel(parent), m_selectionModel(new ItemMultiSelectionModel(this))
+{
+    connect(m_selectionModel, &ItemMultiSelectionModel::selectionChanged, this, &SelectMultipleDirectoriesModel::selectionChanged);
+}
+
+QVariant SelectMultipleDirectoriesModel::data(const QModelIndex& index, int role) const
+{
+    if (!index.isValid()) {
+        return QVariant();
+    }
+
+    io::path path = m_directories[index.row()];
+    switch (role) {
+    case TitleRole: return path.toQString();
+    case SelectedRole: return m_selectionModel->isSelected(index);
+    }
+
+    return QVariant();
+}
+
+bool SelectMultipleDirectoriesModel::setData(const QModelIndex& index, const QVariant& value, int role)
+{
+    if (!index.isValid()) {
+        return false;
+    }
+
+    switch (role) {
+    case TitleRole:
+        m_directories[index.row()] = value.toString().toStdString();
+        emit dataChanged(index, index, { TitleRole });
+        return true;
+    }
+
+    return false;
+}
+
+int SelectMultipleDirectoriesModel::rowCount(const QModelIndex&) const
+{
+    return static_cast<int>(m_directories.size());
+}
+
+QHash<int, QByteArray> SelectMultipleDirectoriesModel::roleNames() const
+{
+    static const QHash<int, QByteArray> roles = {
+        { TitleRole, "titleRole" },
+        { SelectedRole, "selectedRole" }
+    };
+
+    return roles;
+}
+
+QItemSelectionModel* SelectMultipleDirectoriesModel::selection() const
+{
+    return m_selectionModel;
+}
+
+void SelectMultipleDirectoriesModel::load(const QString& dir, const QString& directoriesStr)
+{
+    beginResetModel();
+    m_directories = io::pathsFromString(directoriesStr.toStdString());
+    m_originDirectories = m_directories;
+    m_dir = dir;
+    endResetModel();
+}
+
+void SelectMultipleDirectoriesModel::selectRow(int row)
+{
+    if (!isIndexValid(row)) {
+        return;
+    }
+
+    QModelIndex rowIndex = index(row);
+    m_selectionModel->select(rowIndex);
+    emit dataChanged(index(0), index(rowCount() - 1), { SelectedRole });
+}
+
+void SelectMultipleDirectoriesModel::removeSelectedDirectories()
+{
+    if (!m_selectionModel->hasSelection()) {
+        return;
+    }
+
+    QList<io::path> directoriesToRemove;
+    for (const QModelIndex& index: m_selectionModel->selectedIndexes()) {
+        directoriesToRemove << m_directories[index.row()];
+    }
+
+    for (int i = m_directories.size() - 1; i >= 0; i--) {
```

Comment:
I would suggest rewriting this code as follows:

```
for (const io::path& dirToRemove : directoriesToRemove) {
    int index = m_directories.indexOf(dirToRemove);
    doRemoveDirectory(index);
}
```

This implementation is a bit more reliable